File: core/transcription/src/commonMain/kotlin/dev/audiocompanion/transcription/SonioxTranscriptionProvider.kt
package dev.audiocompanion.transcription

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.delete
import io.ktor.client.request.forms.MultiPartFormDataContent
import io.ktor.client.request.forms.formData
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.Headers
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.ktor.http.isSuccess
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.io.Buffer
import kotlinx.io.readByteArray
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.coroutines.cancellation.CancellationException

/**
 * Cloud transcription provider backed by the Soniox async REST API
 * (`https://api.soniox.com/v1`). A bounded, durable segment is transcribed by:
 *
 *  1. `POST /v1/files` (multipart) -> file id,
 *  2. `POST /v1/transcriptions` (json, references the file) -> transcription id,
 *  3. poll `GET /v1/transcriptions/{id}` until `status` is `completed`/`error`,
 *  4. `GET /v1/transcriptions/{id}/transcript` -> tokens, then best-effort delete of both.
 *
 * The synchronous [transcribe] runs the whole flow (polling included), so it plugs into the
 * existing [TranscriptionProcessor]/[TranscriptionModeRouter] for closed segments with no changes
 * to the local path. Speaker diarization is opt-in ([diarizationEnabled]); per-token speakers are
 * grouped into [TranscriptSegment]s. The detached background-upload variant lives in a later layer;
 * this is the foreground / BGProcessing-window path.
 */
class SonioxTranscriptionProvider(
    private val client: HttpClient,
    private val apiKey: () -> String?,
    private val cloudConsent: () -> Boolean,
    private val diarizationEnabled: () -> Boolean = { false },
    private val languageHints: () -> List<String> = { emptyList() },
    private val contextText: () -> String? = { null },
    private val contextTerms: () -> List<String> = { emptyList() },
    private val model: () -> String = { DEFAULT_MODEL },
    private val baseUrl: String = DEFAULT_BASE_URL,
    private val pollIntervalMs: Long = DEFAULT_POLL_INTERVAL_MS,
    private val maxPollAttempts: Int = DEFAULT_MAX_POLL_ATTEMPTS,
    private val maxUploadBytes: Int = DEFAULT_MAX_UPLOAD_BYTES,
    private val sleep: suspend (Long) -> Unit = { delay(it) },
) : TranscriptionProvider, CloudUploadCapable, CloudConnectivityCheck {
    override val id: String = PROVIDER_ID
    override val status: StateFlow<ProviderStatus> = MutableStateFlow(ProviderStatus.Ready)

    private val json = Json { ignoreUnknownKeys = true }

    override suspend fun isAvailable(): Boolean =
        cloudConsent() && !apiKey().isNullOrBlank()

    override suspend fun checkConnectivity(): CloudConnectivityResult {
        val key = apiKey()?.takeIf { it.isNotBlank() }
            ?: return CloudConnectivityResult.NotConfigured(
                "Add a Soniox API key to use cloud transcription.",
            )
        return try {
            // Cheapest authenticated call: list transcriptions. 2xx = key accepted.
            val response = client.get("$baseUrl/v1/transcriptions") {
                header(HttpHeaders.Authorization, "Bearer $key")
            }
            when {
                response.status.isSuccess() -> CloudConnectivityResult.Ok("Soniox connected")
                response.status == HttpStatusCode.Unauthorized ||
                    response.status == HttpStatusCode.Forbidden ->
                    CloudConnectivityResult.Failed("Soniox rejected the API key.")
                else -> CloudConnectivityResult.Failed(
                    "Soniox check failed (${response.status.value}): " +
                        response.body<String>().take(160),
                )
            }
        } catch (e: CancellationException) {
            throw e
        } catch (t: Throwable) {
            CloudConnectivityResult.Failed("Could not reach Soniox: ${t.message ?: "network error"}")
        }
    }

    override suspend fun transcribe(
        pcmChunks: Flow<ByteArray>,
        sampleRateHz: Int,
    ): TranscriptionResult {
        if (!cloudConsent()) throw TranscriptionException.ProviderUnavailable(id)
        val key = apiKey()?.takeIf { it.isNotBlank() }
            ?: throw TranscriptionException.ProviderUnavailable(id)

        val wav = encodeWav(pcmChunks, sampleRateHz)

        val fileId = uploadFile(key, wav)
        try {
            val transcriptionId = createTranscription(key, fileId)
            try {
                awaitCompletion(key, transcriptionId)
                val transcript = fetchTranscript(key, transcriptionId)
                return mapTranscript(transcript)
            } finally {
                deleteQuietly(key, "$baseUrl/v1/transcriptions/$transcriptionId")
            }
        } finally {
            deleteQuietly(key, "$baseUrl/v1/files/$fileId")
        }
    }

    // --- CloudUploadCapable: background-upload the file, finish the control plane in an awake window

    override suspend fun uploadPlan(wav: ByteArray, sampleRateHz: Int): CloudUploadPlan? {
        if (!cloudConsent()) return null
        val key = apiKey()?.takeIf { it.isNotBlank() } ?: return null
        if (wav.size > maxUploadBytes) return null
        return CloudUploadPlan(
            url = "$baseUrl/v1/files",
            headers = mapOf("Authorization" to "Bearer $key"),
            file = MultipartBody.FilePart("file", "segment.wav", "audio/wav", wav),
        )
    }

    override suspend fun onUploadResponse(httpStatus: Int, body: String): CloudUploadStep {
        if (httpStatus !in 200..299) {
            throw failure("file upload", HttpStatusCode.fromValue(httpStatus), body)
        }
        val fileId = json.decodeFromString(SonioxFile.serializer(), body).id
            ?: throw TranscriptionException.TranscriptionFailed("Soniox upload returned no file id")
        return CloudUploadStep.NeedsControlPlane(fileId)
    }

    override suspend fun completeControlPlane(controlState: String): TranscriptionResult {
        val key = apiKey()?.takeIf { it.isNotBlank() }
            ?: throw TranscriptionException.ProviderUnavailable(id)
        try {
            val transcriptionId = createTranscription(key, controlState)
            try {
                awaitCompletion(key, transcriptionId)
                return mapTranscript(fetchTranscript(key, transcriptionId))
            } finally {
                deleteQuietly(key, "$baseUrl/v1/transcriptions/$transcriptionId")
            }
        } finally {
            deleteQuietly(key, "$baseUrl/v1/files/$controlState")
        }
    }

    private suspend fun encodeWav(pcmChunks: Flow<ByteArray>, sampleRateHz: Int): ByteArray {
        val buffer = Buffer()
        try {
            pcmChunks.collect { chunk ->
                if (chunk.isNotEmpty()) {
                    if (buffer.size + chunk.size > maxUploadBytes.toLong()) {
                        throw TranscriptionException.TranscriptionFailed(
                            "segment audio exceeds $maxUploadBytes byte Soniox upload limit",
                        )
                    }
                    buffer.write(chunk)
                }
            }
            val pcm = buffer.readByteArray(buffer.size.toInt())
            if (pcm.isEmpty()) {
                throw TranscriptionException.NoSpeechDetected("empty audio for Soniox upload")
            }
            return PcmWav.encodeMono16(pcm, sampleRateHz)
        } finally {
            buffer.close()
        }
    }

    private suspend fun uploadFile(key: String, wav: ByteArray): String {
        val response: HttpResponse = client.post("$baseUrl/v1/files") {
            header(HttpHeaders.Authorization, "Bearer $key")
            setBody(
                MultiPartFormDataContent(
                    formData {
                        append(
                            "file",
                            wav,
                            Headers.build {
                                append(HttpHeaders.ContentType, "audio/wav")
                                append(HttpHeaders.ContentDisposition, "filename=\"segment.wav\"")
                            },
                        )
                    },
                ),
            )
        }
        val body = response.body<String>()
        if (!response.status.isSuccess()) {
            throw failure("file upload", response.status, body)
        }
        return json.decodeFromString(SonioxFile.serializer(), body).id
            ?: throw TranscriptionException.TranscriptionFailed("Soniox upload returned no file id")
    }

    private suspend fun createTranscription(key: String, fileId: String): String {
        val request = SonioxCreateTranscription(
            model = model(),
            fileId = fileId,
            enableSpeakerDiarization = diarizationEnabled(),
            languageHints = languageHints().ifEmpty { null },
            context = sonioxContextFrom(contextText, contextTerms),
        )
        val response: HttpResponse = client.post("$baseUrl/v1/transcriptions") {
            header(HttpHeaders.Authorization, "Bearer $key")
            header(HttpHeaders.ContentType, "application/json")
            setBody(json.encodeToString(SonioxCreateTranscription.serializer(), request))
        }
        val body = response.body<String>()
        if (!response.status.isSuccess()) {
            throw failure("create transcription", response.status, body)
        }
        return json.decodeFromString(SonioxTranscription.serializer(), body).id
            ?: throw TranscriptionException.TranscriptionFailed("Soniox create returned no id")
    }

    private suspend fun awaitCompletion(key: String, transcriptionId: String) {
        repeat(maxPollAttempts) { attempt ->
            val response: HttpResponse = client.get("$baseUrl/v1/transcriptions/$transcriptionId") {
                header(HttpHeaders.Authorization, "Bearer $key")
            }
            val body = response.body<String>()
            if (!response.status.isSuccess()) {
                throw failure("poll", response.status, body)
            }
            val transcription = json.decodeFromString(SonioxTranscription.serializer(), body)
            when (transcription.status) {
                STATUS_COMPLETED -> return
                STATUS_ERROR -> throw TranscriptionException.TranscriptionFailed(
                    "Soniox transcription error: ${transcription.errorMessage ?: "unknown"}",
                )
                else -> if (attempt < maxPollAttempts - 1) sleep(pollIntervalMs)
            }
        }
        throw TranscriptionException.TranscriptionFailed("Soniox transcription timed out")
    }

    private suspend fun fetchTranscript(key: String, transcriptionId: String): SonioxTranscript {
        val response: HttpResponse =
            client.get("$baseUrl/v1/transcriptions/$transcriptionId/transcript") {
                header(HttpHeaders.Authorization, "Bearer $key")
            }
        val body = response.body<String>()
        if (!response.status.isSuccess()) {
            throw failure("fetch transcript", response.status, body)
        }
        return json.decodeFromString(SonioxTranscript.serializer(), body)
    }

    private fun mapTranscript(transcript: SonioxTranscript): TranscriptionResult {
        val spokenTokens = transcript.tokens.filter { it.isAudioEvent != true && it.text.isNotBlank() }
        val text = transcript.text?.trim().takeUnless { it.isNullOrBlank() }
            ?: spokenTokens.joinToString(separator = "") { it.text }.trim()
        if (text.isBlank()) {
            throw TranscriptionException.NoSpeechDetected("Soniox returned no speech")
        }
        return TranscriptionResult(
            text = text,
            providerId = id,
            modelUsed = model(),
            segments = groupIntoSegments(spokenTokens),
            words = spokenTokens.map { token ->
                TranscriptWord(
                    text = token.text.trim(),
                    startMs = token.startMs,
                    endMs = token.endMs.coerceAtLeast(token.startMs),
                )
            }.filter { it.text.isNotBlank() },
        )
    }

    /** Groups consecutive tokens into segments, breaking on a speaker change or a silence gap. */
    private fun groupIntoSegments(tokens: List<SonioxToken>): List<TranscriptSegment> {
        if (tokens.isEmpty()) return emptyList()
        val segments = mutableListOf<TranscriptSegment>()
        val current = StringBuilder()
        var start = tokens.first().startMs
        var end = tokens.first().endMs
        var speaker = tokens.first().speaker
        var prevEnd = tokens.first().startMs

        fun flush() {
            val cleaned = current.toString().trim()
            if (cleaned.isNotBlank()) {
                segments += TranscriptSegment(
                    text = cleaned,
                    startMs = start,
                    endMs = end.coerceAtLeast(start),
                    speaker = speaker,
                )
            }
            current.clear()
        }

        tokens.forEach { token ->
            val speakerChanged = token.speaker != speaker
            val longGap = token.startMs - prevEnd > SEGMENT_GAP_MS
            if (current.isNotEmpty() && (speakerChanged || longGap)) {
                flush()
                start = token.startMs
                speaker = token.speaker
            }
            current.append(token.text)
            end = token.endMs
            prevEnd = token.endMs
        }
        flush()
        return segments
    }

    private suspend fun deleteQuietly(key: String, url: String) {
        try {
            client.delete(url) { header(HttpHeaders.Authorization, "Bearer $key") }
        } catch (e: CancellationException) {
            throw e
        } catch (_: Throwable) {
            // Best-effort cleanup; leaving a file/transcription behind is not worth failing over.
        }
    }

    private fun failure(stage: String, status: HttpStatusCode, body: String): TranscriptionException =
        TranscriptionException.TranscriptionFailed(
            "Soniox $stage failed (${status.value}): ${body.take(240)}",
        )

    @Serializable
    private data class SonioxFile(val id: String? = null)

    @Serializable
    private data class SonioxCreateTranscription(
        val model: String,
        @kotlinx.serialization.SerialName("file_id") val fileId: String,
        @kotlinx.serialization.SerialName("enable_speaker_diarization")
        val enableSpeakerDiarization: Boolean = false,
        @kotlinx.serialization.SerialName("language_hints")
        val languageHints: List<String>? = null,
        val context: SonioxContext? = null,
    )

    @Serializable
    private data class SonioxTranscription(
        val id: String? = null,
        val status: String? = null,
        @kotlinx.serialization.SerialName("error_message") val errorMessage: String? = null,
    )

    @Serializable
    private data class SonioxTranscript(
        val text: String? = null,
        val tokens: List<SonioxToken> = emptyList(),
    )

    @Serializable
    private data class SonioxToken(
        val text: String = "",
        @kotlinx.serialization.SerialName("start_ms") val startMs: Long = 0,
        @kotlinx.serialization.SerialName("end_ms") val endMs: Long = 0,
        val speaker: String? = null,
        @kotlinx.serialization.SerialName("is_audio_event") val isAudioEvent: Boolean? = null,
    )

    companion object {
        const val PROVIDER_ID = "soniox"
        const val DEFAULT_BASE_URL = "https://api.soniox.com"
        const val DEFAULT_MODEL = "stt-async-v5"
        private const val DEFAULT_POLL_INTERVAL_MS = 2_000L
        private const val DEFAULT_MAX_POLL_ATTEMPTS = 150
        private const val DEFAULT_MAX_UPLOAD_BYTES = 100 * 1024 * 1024
        private const val SEGMENT_GAP_MS = 800L
        private const val STATUS_COMPLETED = "completed"
        private const val STATUS_ERROR = "error"
    }
}

File: ios/PebbleAudioKit/Sources/Transcription/Providers/SonioxTranscriptionProvider.swift
import AudioCodec
import Foundation

// Port of `core/transcription/.../SonioxTranscriptionProvider.kt`.

/// Cloud transcription provider backed by the Soniox async REST API
/// (`https://api.soniox.com/v1`). A bounded, durable segment is transcribed by:
///
///  1. `POST /v1/files` (multipart) -> file id,
///  2. `POST /v1/transcriptions` (json, references the file) -> transcription id,
///  3. poll `GET /v1/transcriptions/{id}` until `status` is `completed`/`error`,
///  4. `GET /v1/transcriptions/{id}/transcript` -> tokens, then best-effort delete of both.
///
/// The synchronous `transcribe` runs the whole flow (polling included), so it plugs into the
/// processor/router for closed segments with no changes to the local path. Speaker diarization is
/// opt-in (`diarizationEnabled`); per-token speakers are grouped into `TranscriptSegment`s. The
/// detached background-upload variant is the `CloudUploadCapable` conformance below.
public final class SonioxTranscriptionProvider: TranscriptionProvider, CloudUploadCapable {
    public static let providerId = "soniox"
    public static let defaultBaseUrl = "https://api.soniox.com"
    public static let defaultModel = "stt-async-v5"
    public static let defaultPollIntervalMs: Int64 = 2_000
    public static let defaultMaxPollAttempts = 150
    public static let defaultMaxUploadBytes = 100 * 1024 * 1024
    static let segmentGapMs: Int64 = 800
    private static let statusCompleted = "completed"
    private static let statusError = "error"

    public let id = SonioxTranscriptionProvider.providerId

    private let transport: any HttpTransport
    private let apiKey: @Sendable () -> String?
    private let cloudConsent: @Sendable () -> Bool
    private let diarizationEnabled: @Sendable () -> Bool
    private let languageHints: @Sendable () -> [String]
    private let contextText: @Sendable () -> String?
    private let contextTerms: @Sendable () -> [String]
    private let model: @Sendable () -> String
    let baseUrl: String
    private let pollIntervalMs: Int64
    private let maxPollAttempts: Int
    private let maxUploadBytes: Int
    /// Injectable so tests can poll without real delays.
    private let sleep: @Sendable (Int64) async throws -> Void

    public init(
        transport: any HttpTransport,
        apiKey: @escaping @Sendable () -> String?,
        cloudConsent: @escaping @Sendable () -> Bool,
        diarizationEnabled: @escaping @Sendable () -> Bool = { false },
        languageHints: @escaping @Sendable () -> [String] = { [] },
        contextText: @escaping @Sendable () -> String? = { nil },
        contextTerms: @escaping @Sendable () -> [String] = { [] },
        model: @escaping @Sendable () -> String = { SonioxTranscriptionProvider.defaultModel },
        baseUrl: String = SonioxTranscriptionProvider.defaultBaseUrl,
        pollIntervalMs: Int64 = SonioxTranscriptionProvider.defaultPollIntervalMs,
        maxPollAttempts: Int = SonioxTranscriptionProvider.defaultMaxPollAttempts,
        maxUploadBytes: Int = SonioxTranscriptionProvider.defaultMaxUploadBytes,
        sleep: @escaping @Sendable (Int64) async throws -> Void = { ms in
            try await Task.sleep(nanoseconds: UInt64(max(0, ms)) * 1_000_000)
        }
    ) {
        self.transport = transport
        self.apiKey = apiKey
        self.cloudConsent = cloudConsent
        self.diarizationEnabled = diarizationEnabled
        self.languageHints = languageHints
        self.contextText = contextText
        self.contextTerms = contextTerms
        self.model = model
        self.baseUrl = baseUrl
        self.pollIntervalMs = pollIntervalMs
        self.maxPollAttempts = maxPollAttempts
        self.maxUploadBytes = maxUploadBytes
        self.sleep = sleep
    }

    public func isAvailable() async -> Bool {
        cloudConsent() && !(apiKey() ?? "").isBlank
    }

    // Connectivity-probe seam (see CloudConnectivityChecks.swift).
    var connectivityTransport: any HttpTransport { transport }
    var connectivityApiKey: String? { apiKey() }

    public func transcribe(
        pcmChunks: AsyncThrowingStream<Data, Error>,
        sampleRateHz: Int
    ) async throws -> TranscriptionResult {
        guard cloudConsent() else { throw TranscriptionError.providerUnavailable(providerId: id) }
        guard let key = apiKey()?.nonBlank else {
            throw TranscriptionError.providerUnavailable(providerId: id)
        }

        let wav = try await encodeWav(pcmChunks: pcmChunks, sampleRateHz: sampleRateHz)

        let fileId = try await uploadFile(key: key, wav: wav)
        do {
            let result = try await transcribeUploadedFile(key: key, fileId: fileId)
            await deleteQuietly(key: key, url: "\(baseUrl)/v1/files/\(fileId)")
            return result
        } catch {
            await deleteQuietly(key: key, url: "\(baseUrl)/v1/files/\(fileId)")
            throw error
        }
    }

    /// Create -> poll -> fetch -> map for an already-uploaded file; always best-effort deletes
    /// the transcription record (the caller owns deleting the file).
    private func transcribeUploadedFile(key: String, fileId: String) async throws -> TranscriptionResult {
        let transcriptionId = try await createTranscription(key: key, fileId: fileId)
        do {
            try await awaitCompletion(key: key, transcriptionId: transcriptionId)
            let transcript = try await fetchTranscript(key: key, transcriptionId: transcriptionId)
            let result = try mapTranscript(transcript)
            await deleteQuietly(key: key, url: "\(baseUrl)/v1/transcriptions/\(transcriptionId)")
            return result
        } catch {
            await deleteQuietly(key: key, url: "\(baseUrl)/v1/transcriptions/\(transcriptionId)")
            throw error
        }
    }

    // MARK: - CloudUploadCapable: background-upload the file, finish the control plane when awake

    public func uploadPlan(wav: Data, sampleRateHz: Int) async -> CloudUploadPlan? {
        guard cloudConsent(), let key = apiKey()?.nonBlank else { return nil }
        if wav.count > maxUploadBytes { return nil }
        return CloudUploadPlan(
            url: "\(baseUrl)/v1/files",
            headers: ["Authorization": "Bearer \(key)"],
            textFields: [],
            file: MultipartBody.FilePart(
                name: "file", filename: "segment.wav", contentType: "audio/wav", bytes: wav
            )
        )
    }

    public func onUploadResponse(httpStatus: Int, body: String) async throws -> CloudUploadStep {
        guard (200..<300).contains(httpStatus) else {
            throw failure(stage: "file upload", status: httpStatus, body: body)
        }
        guard let fileId = try? JSONDecoder().decode(SonioxFile.self, from: Data(body.utf8)).id else {
            throw TranscriptionError.transcriptionFailed("Soniox upload returned no file id")
        }
        return .needsControlPlane(fileId)
    }

    public func completeControlPlane(controlState: String) async throws -> TranscriptionResult {
        guard let key = apiKey()?.nonBlank else {
            throw TranscriptionError.providerUnavailable(providerId: id)
        }
        do {
            let result = try await transcribeUploadedFile(key: key, fileId: controlState)
            await deleteQuietly(key: key, url: "\(baseUrl)/v1/files/\(controlState)")
            return result
        } catch {
            await deleteQuietly(key: key, url: "\(baseUrl)/v1/files/\(controlState)")
            throw error
        }
    }

    // MARK: - Flow steps

    private func encodeWav(
        pcmChunks: AsyncThrowingStream<Data, Error>,
        sampleRateHz: Int
    ) async throws -> Data {
        var pcm = Data()
        for try await chunk in pcmChunks where !chunk.isEmpty {
            if pcm.count + chunk.count > maxUploadBytes {
                throw TranscriptionError.transcriptionFailed(
                    "segment audio exceeds \(maxUploadBytes) byte Soniox upload limit"
                )
            }
            pcm.append(chunk)
        }
        if pcm.isEmpty {
            throw TranscriptionError.noSpeechDetected("empty audio for Soniox upload")
        }
        return PcmWav.encodeMono16(pcm: pcm, sampleRateHz: sampleRateHz)
    }

    private func uploadFile(key: String, wav: Data) async throws -> String {
        let boundary = "PebbleAudioKit-\(UUID().uuidString)"
        let (body, contentType) = MultipartBody.encode(
            boundary: boundary,
            textFields: [],
            file: MultipartBody.FilePart(
                name: "file", filename: "segment.wav", contentType: "audio/wav", bytes: wav
            )
        )
        let response = try await transport.execute(
            HttpTransportRequest(
                method: "POST",
                url: "\(baseUrl)/v1/files",
                headers: [
                    "Authorization": "Bearer \(key)",
                    "Content-Type": contentType,
                ],
                body: body
            )
        )
        guard response.isSuccess else {
            throw failure(stage: "file upload", status: response.status, body: response.text)
        }
        guard let id = try? JSONDecoder().decode(SonioxFile.self, from: response.body).id else {
            throw TranscriptionError.transcriptionFailed("Soniox upload returned no file id")
        }
        return id
    }

    private func createTranscription(key: String, fileId: String) async throws -> String {
        let request = CreateTranscription(
            model: model(),
            fileId: fileId,
            enableSpeakerDiarization: diarizationEnabled(),
            languageHints: languageHints().isEmpty ? nil : languageHints(),
            context: sonioxContextFrom(contextText: contextText, contextTerms: contextTerms)
        )
        let body = try JSONEncoder().encode(request)
        let response = try await transport.execute(
            HttpTransportRequest(
                method: "POST",
                url: "\(baseUrl)/v1/transcriptions",
                headers: [
                    "Authorization": "Bearer \(key)",
                    "Content-Type": "application/json",
                ],
                body: body
            )
        )
        guard response.isSuccess else {
            throw failure(stage: "create transcription", status: response.status, body: response.text)
        }
        guard let id = try? JSONDecoder().decode(SonioxTranscriptionMeta.self, from: response.body).id
        else {
            throw TranscriptionError.transcriptionFailed("Soniox create returned no id")
        }
        return id
    }

    private func awaitCompletion(key: String, transcriptionId: String) async throws {
        for attempt in 0..<maxPollAttempts {
            let response = try await transport.execute(
                HttpTransportRequest(
                    method: "GET",
                    url: "\(baseUrl)/v1/transcriptions/\(transcriptionId)",
                    headers: ["Authorization": "Bearer \(key)"]
                )
            )
            guard response.isSuccess else {
                throw failure(stage: "poll", status: response.status, body: response.text)
            }
            let meta = try JSONDecoder().decode(SonioxTranscriptionMeta.self, from: response.body)
            switch meta.status {
            case Self.statusCompleted:
                return
            case Self.statusError:
                throw TranscriptionError.transcriptionFailed(
                    "Soniox transcription error: \(meta.errorMessage ?? "unknown")"
                )
            default:
                if attempt < maxPollAttempts - 1 { try await sleep(pollIntervalMs) }
            }
        }
        throw TranscriptionError.transcriptionFailed("Soniox transcription timed out")
    }

    private func fetchTranscript(key: String, transcriptionId: String) async throws -> SonioxTranscript {
        let response = try await transport.execute(
            HttpTransportRequest(
                method: "GET",
                url: "\(baseUrl)/v1/transcriptions/\(transcriptionId)/transcript",
                headers: ["Authorization": "Bearer \(key)"]
            )
        )
        guard response.isSuccess else {
            throw failure(stage: "fetch transcript", status: response.status, body: response.text)
        }
        return try JSONDecoder().decode(SonioxTranscript.self, from: response.body)
    }

    private func mapTranscript(_ transcript: SonioxTranscript) throws -> TranscriptionResult {
        let spokenTokens = (transcript.tokens ?? []).filter {
            $0.isAudioEvent != true && !($0.text ?? "").isBlank
        }
        let text = transcript.text?.trimmed.nonBlank
            ?? spokenTokens.map { $0.text ?? "" }.joined().trimmed
        if text.isBlank {
            throw TranscriptionError.noSpeechDetected("Soniox returned no speech")
        }
        return TranscriptionResult(
            text: text,
            providerId: id,
            modelUsed: model(),
            segments: groupIntoSegments(spokenTokens),
            words: spokenTokens.compactMap { token in
                let cleaned = (token.text ?? "").trimmed
                if cleaned.isEmpty { return nil }
                let startMs = token.startMs ?? 0
                return TranscriptWord(
                    text: cleaned,
                    startMs: startMs,
                    endMs: max(token.endMs ?? 0, startMs)
                )
            }
        )
    }

    /// Groups consecutive tokens into segments, breaking on a speaker change or a silence gap.
    private func groupIntoSegments(_ tokens: [SonioxBatchToken]) -> [TranscriptSegment] {
        guard let first = tokens.first else { return [] }
        var segments: [TranscriptSegment] = []
        var current = ""
        var start = first.startMs ?? 0
        var end = first.endMs ?? 0
        var speaker = first.speaker
        var prevEnd = first.startMs ?? 0

        func flush() {
            let cleaned = current.trimmed
            if !cleaned.isEmpty {
                segments.append(
                    TranscriptSegment(
                        text: cleaned, startMs: start, endMs: max(end, start), speaker: speaker
                    )
                )
            }
            current = ""
        }

        for token in tokens {
            let tokenStart = token.startMs ?? 0
            let speakerChanged = token.speaker != speaker
            let longGap = tokenStart - prevEnd > Self.segmentGapMs
            if !current.isEmpty && (speakerChanged || longGap) {
                flush()
                start = tokenStart
                speaker = token.speaker
            }
            current += token.text ?? ""
            end = token.endMs ?? 0
            prevEnd = token.endMs ?? 0
        }
        flush()
        return segments
    }

    /// Best-effort cleanup; leaving a file/transcription behind is not worth failing over.
    private func deleteQuietly(key: String, url: String) async {
        _ = try? await transport.execute(
            HttpTransportRequest(
                method: "DELETE", url: url, headers: ["Authorization": "Bearer \(key)"]
            )
        )
    }

    private func failure(stage: String, status: Int, body: String) -> TranscriptionError {
        .transcriptionFailed("Soniox \(stage) failed (\(status)): \(body.prefix(240))")
    }

    // MARK: - Wire shapes

    private struct SonioxFile: Decodable {
        var id: String?
    }

    private struct CreateTranscription: Encodable {
        var model: String
        var fileId: String
        var enableSpeakerDiarization: Bool
        var languageHints: [String]?
        var context: SonioxContext?

        enum CodingKeys: String, CodingKey {
            case model
            case fileId = "file_id"
            case enableSpeakerDiarization = "enable_speaker_diarization"
            case languageHints = "language_hints"
            case context
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.container(keyedBy: CodingKeys.self)
            try container.encode(model, forKey: .model)
            try container.encode(fileId, forKey: .fileId)
            // kotlinx.serialization omits default values: false is the default, so only encode
            // the flag when diarization is on (keeps the wire body byte-compatible).
            if enableSpeakerDiarization {
                try container.encode(true, forKey: .enableSpeakerDiarization)
            }
            try container.encodeIfPresent(languageHints, forKey: .languageHints)
            try container.encodeIfPresent(context, forKey: .context)
        }
    }

    private struct SonioxTranscriptionMeta: Decodable {
        var id: String?
        var status: String?
        var errorMessage: String?

        enum CodingKeys: String, CodingKey {
            case id
            case status
            case errorMessage = "error_message"
        }
    }

    struct SonioxTranscript: Decodable {
        var text: String?
        var tokens: [SonioxBatchToken]?
    }

    struct SonioxBatchToken: Decodable {
        var text: String?
        var startMs: Int64?
        var endMs: Int64?
        var speaker: String?
        var isAudioEvent: Bool?

        enum CodingKeys: String, CodingKey {
            case text
            case startMs = "start_ms"
            case endMs = "end_ms"
            case speaker
            case isAudioEvent = "is_audio_event"
        }
    }
}

extension String {
    fileprivate var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
    fileprivate var isBlank: Bool { trimmed.isEmpty }
    fileprivate var nonBlank: String? { isBlank ? nil : self }
}
